### PR TITLE
Display conflict-free merge messages for pull requests

### PR DESCRIPTION
**Pull request by justusbunsi**
External issue tracker often use a syntax like `#1337` as item references in commits. When you handle pull requests with the default "merge commit" message, the very pull request will be referenced with `#` as well. Gitea itself can handle this situation and is able to identify both item and pull request reference when parsing the commit message.
If other external components (e.g.) push hooks ensure commits to only contain valid item references by looking for `#` occurrences, such commits would be rejected due to a non-valid issue reference.

Regarding the [documentation](https://docs.gitea.io/en-us/automatically-linked-references/#issues-and-pull-requests), Gitea is able to use `#` and `!` for item references for these use cases. This was handled only for squash merges

These changes take an external issue tracker into account when rendering merge commit messages. In such a case the documented `!` will be used for item references in merge commits.

Signed-off-by: Steven Kriegler <61625851+justusbunsi@users.noreply.github.com>

### CODE REVIEW

**Review by lafriks — APPROVED** [review]

File: models/pull.go [hunk 1]
```diff
@@ -212,12 +212,21 @@ func (pr *PullRequest) GetDefaultMergeMessage() string {
...
 	}
+	if err := pr.LoadBaseRepo(); err != nil {
+		log.Error("LoadBaseRepo: %v", err)
+		return ""
+	}
+
+	issueReference := fmt.Sprintf("#%d", pr.Issue.Index)
+	if pr.BaseRepo.UnitEnabled(UnitTypeExternalTracker) {
+		issueReference = fmt.Sprintf("!%d", pr.Issue.Index)
+	}
```
Comment (zeripath) [inline models/pull.go]:
```suggestion
	issueReference := "#"
	if pr.BaseRepo.UnitEnabled(UnitTypeExternalTracker) {
		issueReference = "!"
	}
```

File: models/pull.go [hunk 2]
```diff
@@ -212,12 +212,21 @@ func (pr *PullRequest) GetDefaultMergeMessage() string {
...
 
 	if pr.BaseRepoID == pr.HeadRepoID {
-		return fmt.Sprintf("Merge pull request '%s' (#%d) from %s into %s", pr.Issue.Title, pr.Issue.Index, pr.HeadBranch, pr.BaseBranch)
+		return fmt.Sprintf("Merge pull request '%s' (%s) from %s into %s", pr.Issue.Title, issueReference, pr.HeadBranch, pr.BaseBranch)
```
Comment (zeripath) [inline models/pull.go]:
```suggestion
		return fmt.Sprintf("Merge pull request '%s' (%s%d) from %s into %s", pr.Issue.Title, issueReference, pr.Issue.Index, pr.HeadBranch, pr.BaseBranch)
```

File: models/pull.go [hunk 3]
```diff
@@ -212,12 +212,21 @@ func (pr *PullRequest) GetDefaultMergeMessage() string {
...
 	}
 
-	return fmt.Sprintf("Merge pull request '%s' (#%d) from %s:%s into %s", pr.Issue.Title, pr.Issue.Index, pr.HeadRepo.FullName(), pr.HeadBranch, pr.BaseBranch)
+	return fmt.Sprintf("Merge pull request '%s' (%s) from %s:%s into %s", pr.Issue.Title, issueReference, pr.HeadRepo.FullName(), pr.HeadBranch, pr.BaseBranch)
```
Comment (zeripath) [inline models/pull.go]:
```suggestion
	return fmt.Sprintf("Merge pull request '%s' (%s%d) from %s:%s into %s", pr.Issue.Title, issueReference, pr.Issue.Index, pr.HeadRepo.FullName(), pr.HeadBranch, pr.BaseBranch)
```

**Comment by zeripath** [comment]
@wULLSnpAXbWZGYDYyhWTKKspEQoaYxXyhoisqHf I suspect that this needs a second check to work for your situation?

**Comment by wULLSnpAXbWZGYDYyhWTKKspEQoaYxXyhoisqHf** [comment]
> @wULLSnpAXbWZGYDYyhWTKKspEQoaYxXyhoisqHf I suspect that this needs a second check to work for your situation?

Thanks for bringing me in!
Currently building with this patch.
How exactly would I be supposed to check, though? I should probably give it a proper look 🙈 Give me a minute, please.

**Comment by zeripath** [comment]
Actually @wULLSnpAXbWZGYDYyhWTKKspEQoaYxXyhoisqHf you wouldn't use this - so I think you're fine. Sorry for wasting your time.

**Review by zeripath — APPROVED** [review]

**Pull request merged**